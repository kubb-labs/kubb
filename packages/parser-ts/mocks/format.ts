import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string): Promise<string> {
  if (!source) {
    return Promise.resolve('')
  }
  return prettierFormat(source, formatOptions)
}
