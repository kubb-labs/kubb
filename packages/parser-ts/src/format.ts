import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}

/**
 * Format the generated code based on Prettier
 */
export async function format(source: string) {
  // do some basic linting with the ts compiler
  if (!source) {
    return ''
  }

  return prettierFormat(source, formatOptions)
}
