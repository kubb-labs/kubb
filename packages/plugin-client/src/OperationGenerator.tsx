import { createReactParser, OperationGenerator as Generator } from '@kubb/plugin-oas'
import { Oas } from '@kubb/plugin-oas/components'
import { App, createRoot } from '@kubb/react'

import { Client, Operations } from './components/index.ts'

import type { Operation } from '@kubb/oas'
import type { OperationMethodResult, OperationsByMethod } from '@kubb/plugin-oas'
import type { FileMeta, PluginClient } from './types.ts'

export const clientParser = createReactParser<PluginClient>({
  name: 'plugin-client',
  Operations({ options }) {
    if (!options.templates.operations) {
      return null
    }

    return <Operations.File baseURL={options.baseURL} templates={options.templates.operations} />
  },
  Operation({ options, operation }) {
    const isEnabled = options.client.methods.some((method) => operation.method === method)

    if (!options.templates.client || !isEnabled) {
      return null
    }

    return <Client.File baseURL={options.baseURL} templates={options.templates.client} />
  },
})
