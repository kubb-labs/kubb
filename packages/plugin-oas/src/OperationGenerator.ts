import { BaseGenerator, type FileMetaBase } from '@kubb/core'
import transformers from '@kubb/core/transformers'

import type { PluginFactoryOptions, PluginManager } from '@kubb/core'
import type * as KubbFile from '@kubb/fs/types'

import type { Plugin } from '@kubb/core'
import type { HttpMethod, Oas, OasTypes, Operation, contentType } from '@kubb/oas'
import type { Generator } from './generator.tsx'
import type { Exclude, Include, OperationSchemas, OperationsByMethod, Override } from './types.ts'

/**
 * @deprecated
 */
export type GetOperationGeneratorOptions<T extends OperationGenerator<any, any, any>> = T extends OperationGenerator<infer Options, any, any> ? Options : never

export type OperationMethodResult<TFileMeta extends FileMetaBase> = Promise<KubbFile.File<TFileMeta> | Array<KubbFile.File<TFileMeta>> | null>

type Context<TOptions, TPluginOptions extends PluginFactoryOptions> = {
  oas: Oas
  exclude: Array<Exclude> | undefined
  include: Array<Include> | undefined
  override: Array<Override<TOptions>> | undefined
  contentType: contentType | undefined
  pluginManager: PluginManager
  /**
   * Current plugin
   */
  plugin: Plugin<TPluginOptions>
  mode: KubbFile.Mode
}

export class OperationGenerator<
  TOptions = unknown,
  TPluginOptions extends PluginFactoryOptions = PluginFactoryOptions,
  TFileMeta extends FileMetaBase = FileMetaBase,
> extends BaseGenerator<TOptions, Context<TOptions, TPluginOptions>> {
  #operationsByMethod: OperationsByMethod = {}
  get operationsByMethod(): OperationsByMethod {
    return this.#operationsByMethod
  }

  set operationsByMethod(paths: OperationsByMethod) {
    this.#operationsByMethod = paths
  }

  #getOptions(operation: Operation, method: HttpMethod): Partial<TOptions> {
    const { override = [] } = this.context

    return (
      override.find(({ pattern, type }) => {
        if (type === 'tag') {
          return !!operation.getTags()[0]?.name.match(pattern)
        }

        if (type === 'operationId') {
          return !!operation.getOperationId().match(pattern)
        }

        if (type === 'path') {
          return !!operation.path.match(pattern)
        }

        if (type === 'method') {
          return !!method.match(pattern)
        }

        return false
      })?.options || {}
    )
  }

  /**
   *
   * @deprecated
   */
  #isExcluded(operation: Operation, method: HttpMethod): boolean {
    const { exclude = [] } = this.context
    let matched = false

    exclude.forEach(({ pattern, type }) => {
      if (type === 'tag' && !matched) {
        matched = !!operation.getTags()[0]?.name.match(pattern)
      }

      if (type === 'operationId' && !matched) {
        matched = !!operation.getOperationId().match(pattern)
      }

      if (type === 'path' && !matched) {
        matched = !!operation.path.match(pattern)
      }

      if (type === 'method' && !matched) {
        matched = !!method.match(pattern)
      }
    })

    return matched
  }

  /**
   *
   * @deprecated
   */
  #isIncluded(operation: Operation, method: HttpMethod): boolean {
    const { include = [] } = this.context
    let matched = false

    include.forEach(({ pattern, type }) => {
      if (type === 'tag' && !matched) {
        matched = !!operation.getTags()[0]?.name.match(pattern)
      }

      if (type === 'operationId' && !matched) {
        matched = !!operation.getOperationId().match(pattern)
      }

      if (type === 'path' && !matched) {
        matched = !!operation.path.match(pattern)
      }

      if (type === 'method' && !matched) {
        matched = !!method.match(pattern)
      }
    })

    return matched
  }

  getSchemas(
    operation: Operation,
    {
      forStatusCode,
      resolveName = (name) => name,
    }: {
      forStatusCode?: string | number
      resolveName?: (name: string) => string
    } = {},
  ): OperationSchemas {
    const pathParamsSchema = this.context.oas.getParametersSchema(operation, 'path')
    const queryParamsSchema = this.context.oas.getParametersSchema(operation, 'query')
    const headerParamsSchema = this.context.oas.getParametersSchema(operation, 'header')
    const requestSchema = this.context.oas.getRequestSchema(operation)
    const responseStatusCode =
      forStatusCode || (operation.schema.responses && Object.keys(operation.schema.responses).find((key) => key.startsWith('2'))) || 200
    const responseSchema = this.context.oas.getResponseSchema(operation, responseStatusCode)
    const statusCodes = operation.getResponseStatusCodes().map((statusCode) => {
      let name = statusCode
      if (name === 'default') {
        name = 'error'
      }

      const schema = this.context.oas.getResponseSchema(operation, statusCode)

      return {
        name: resolveName(transformers.pascalCase(`${operation.getOperationId()} ${name}`)),
        description: (operation.getResponseByStatusCode(statusCode) as OasTypes.ResponseObject)?.description,
        schema,
        operation,
        operationName: transformers.pascalCase(`${operation.getOperationId()}`),
        statusCode: name === 'error' ? undefined : Number(statusCode),
        keys: schema?.properties ? Object.keys(schema.properties) : undefined,
      }
    })

    return {
      pathParams: pathParamsSchema
        ? {
            name: resolveName(transformers.pascalCase(`${operation.getOperationId()} PathParams`)),
            operation,
            operationName: transformers.pascalCase(`${operation.getOperationId()}`),
            schema: pathParamsSchema,
            keys: pathParamsSchema.properties ? Object.keys(pathParamsSchema.properties) : undefined,
          }
        : undefined,
      queryParams: queryParamsSchema
        ? {
            name: resolveName(transformers.pascalCase(`${operation.getOperationId()} QueryParams`)),
            operation,
            operationName: transformers.pascalCase(`${operation.getOperationId()}`),
            schema: queryParamsSchema,
            keys: queryParamsSchema.properties ? Object.keys(queryParamsSchema.properties) : [],
          }
        : undefined,
      headerParams: headerParamsSchema
        ? {
            name: resolveName(transformers.pascalCase(`${operation.getOperationId()} HeaderParams`)),
            operation,
            operationName: transformers.pascalCase(`${operation.getOperationId()}`),
            schema: headerParamsSchema,
            keys: headerParamsSchema.properties ? Object.keys(headerParamsSchema.properties) : undefined,
          }
        : undefined,
      request: requestSchema
        ? {
            name: resolveName(transformers.pascalCase(`${operation.getOperationId()} ${operation.method === 'get' ? 'queryRequest' : 'mutationRequest'}`)),
            description: (operation.schema.requestBody as OasTypes.RequestBodyObject)?.description,
            operation,
            operationName: transformers.pascalCase(`${operation.getOperationId()}`),
            schema: requestSchema,
            keys: requestSchema.properties ? Object.keys(requestSchema.properties) : undefined,
            keysToOmit: requestSchema.properties
              ? Object.keys(requestSchema.properties).filter((key) => {
                  const item = requestSchema.properties?.[key] as OasTypes.SchemaObject

                  return item?.readOnly
                })
              : undefined,
          }
        : undefined,
      response: {
        name: resolveName(transformers.pascalCase(`${operation.getOperationId()} ${operation.method === 'get' ? 'queryResponse' : 'mutationResponse'}`)),
        description: operation.getResponseAsJSONSchema(responseStatusCode)?.at(0)?.description,
        operation,
        operationName: transformers.pascalCase(`${operation.getOperationId()}`),
        schema: responseSchema,
        statusCode: Number(responseStatusCode),
        keys: responseSchema?.properties ? Object.keys(responseSchema.properties) : undefined,
        keysToOmit: responseSchema?.properties
          ? Object.keys(responseSchema.properties).filter((key) => {
              const item = responseSchema.properties?.[key] as OasTypes.SchemaObject
              return item?.writeOnly
            })
          : undefined,
      },
      errors: statusCodes.filter((item) => item.statusCode?.toString().startsWith('4') || item.statusCode?.toString().startsWith('5')),
      statusCodes,
    }
  }

  #methods = ['get', 'post', 'patch', 'put', 'delete']

  async build(...generators: Array<Generator<Extract<TOptions, PluginFactoryOptions>>>): Promise<Array<KubbFile.File<TFileMeta>>> {
    const { oas } = this.context

    const paths = oas.getPaths()
    this.operationsByMethod = Object.entries(paths).reduce((acc, [path, method]) => {
      const methods = Object.keys(method) as HttpMethod[]

      methods.forEach((method) => {
        const operation = oas.operation(path, method)
        if (operation && [this.#methods].some((methods) => method === operation.method)) {
          const isExcluded = this.#isExcluded(operation, method)
          const isIncluded = this.context.include ? this.#isIncluded(operation, method) : true

          if (isIncluded && !isExcluded) {
            if (!acc[path]) {
              acc[path] = {} as OperationsByMethod['get']
            }
            acc[path] = {
              ...acc[path],
              [method]: {
                operation,
                schemas: this.getSchemas(operation),
              },
            } as OperationsByMethod['get']
          }
        }
      })

      return acc
    }, {} as OperationsByMethod)

    const promises = Object.keys(this.operationsByMethod).reduce((acc, path) => {
      const methods = this.operationsByMethod[path] ? (Object.keys(this.operationsByMethod[path]!) as HttpMethod[]) : []

      methods.forEach((method) => {
        const { operation } = this.operationsByMethod[path]?.[method]!
        const options = this.#getOptions(operation, method)

        const methodToCall = this[method as keyof typeof this] as any

        if (typeof methodToCall === 'function') {
          const promiseMethod = methodToCall?.call(this, operation, {
            ...this.options,
            ...options,
          })

          if (promiseMethod) {
            acc.push(promiseMethod)
          }
        }

        const promiseOperation = this.operation.call(this, operation, {
          ...this.options,
          ...options,
        })

        if (promiseOperation) {
          acc.push(promiseOperation)
        }

        generators?.forEach((generator) => {
          const promise = generator.operation?.({
            instance: this,
            operation,
            options: {
              ...this.options,
              ...options,
            },
          } as any) as Promise<Array<KubbFile.File<TFileMeta>>>

          if (promise) {
            acc.push(promise)
          }
        })
      })

      return acc
    }, [] as OperationMethodResult<TFileMeta>[])

    const operations = Object.values(this.operationsByMethod).map((item) => Object.values(item).map((item) => item.operation))

    promises.push(this.all(operations.flat().filter(Boolean), this.operationsByMethod))

    generators?.forEach((generator) => {
      const promise = generator.operations?.({
        instance: this,
        operations: operations.flat().filter(Boolean),
        operationsByMethod: this.operationsByMethod,
        options: this.options,
      } as any) as Promise<Array<KubbFile.File<TFileMeta>>>

      if (promise) {
        promises.push(promise)
      }
    })

    const files = await Promise.all(promises)

    // using .flat because operationGenerator[method] can return a array of files or just one file
    return files.flat().filter(Boolean)
  }

  /**
   * Operation
   */
  async operation(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * GET
   */
  async get(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * POST
   */
  async post(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }
  /**
   * PATCH
   */
  async patch(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * PUT
   */
  async put(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * DELETE
   */
  async delete(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * Combination of GET, POST, PATCH, PUT, DELETE
   */
  async all(operations: Operation[], paths: OperationsByMethod): OperationMethodResult<TFileMeta> {
    return []
  }
}
