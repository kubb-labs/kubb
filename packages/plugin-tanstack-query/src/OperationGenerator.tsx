import { OperationGenerator as Generator } from '@kubb/plugin-oas'
import { Oas } from '@kubb/plugin-oas/components'
import { App, createRoot } from '@kubb/react'

import { Mutation } from './components/Mutation.tsx'
import { Query } from './components/Query.tsx'
import { QueryKey } from './components/QueryKey.tsx'
import { QueryOptions } from './components/QueryOptions.tsx'

import type { Operation } from '@kubb/oas'
import type { OperationMethodResult } from '@kubb/plugin-oas'
import { QueryImports } from './components/QueryImports.tsx'
import type { FileMeta, PluginTanstackQuery } from './types.ts'

export class OperationGenerator extends Generator<PluginTanstackQuery, FileMeta> {
  async operation(operation: Operation, options: PluginTanstackQuery['resolvedOptions']): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin, mode } = this.context

    const root = createRoot({
      logger: pluginManager.logger,
    })

    const templates = {
      mutation: Mutation.templates,
      query: Query.templates,
      queryOptions: QueryOptions.templates,
      queryKey: QueryKey.templates,
      queryImports: QueryImports.templates,
      ...options.templates,
    }

    const isQuery = typeof options.query === 'boolean' ? true : options.query?.methods.some((method) => operation.method === method)
    const isMutate = !isQuery && options.mutate && options.mutate.methods.some((method) => operation.method === method)

    root.render(
      <App pluginManager={pluginManager} plugin={{ ...plugin, options }} mode={mode}>
        <Oas oas={oas} operations={[operation]} generator={this}>
          <Oas.Operation operation={operation}>
            {isMutate && templates?.mutation && <Mutation.File templates={templates.mutation} />}
            {isQuery && templates?.query && templates.queryKey && templates.queryOptions && templates.queryImports && (
              <Query.File
                templates={{
                  query: templates.query,
                  queryKey: templates.queryKey,
                  queryOptions: templates.queryOptions,
                  queryImports: templates.queryImports,
                }}
              />
            )}
          </Oas.Operation>
        </Oas>
      </App>,
    )

    return root.files
  }
}
