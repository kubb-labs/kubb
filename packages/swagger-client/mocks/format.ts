import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string) {
  if (!source) {
    return ''
  }
  return prettierFormat(source, formatOptions)
}
