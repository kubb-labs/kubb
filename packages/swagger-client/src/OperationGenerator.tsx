import { createRoot } from '@kubb/react'
import { OperationGenerator as Generator } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { Client, Operations } from './components/index.ts'

import type { AppContextProps } from '@kubb/react'
import type { OperationMethodResult, OperationsByMethod } from '@kubb/swagger'
import type { Operation } from '@kubb/swagger/oas'
import type { FileMeta, PluginOptions } from './types.ts'

export class OperationGenerator extends Generator<PluginOptions['resolvedOptions'], PluginOptions> {
  async all(operations: Operation[], _operationsByMethod: OperationsByMethod): OperationMethodResult<FileMeta> {
    const { pluginManager, oas, plugin } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })

    const templates = {
      operations: Operations.templates,
      client: Client.templates,
      ...this.options.templates,
    }

    root.render(
      <Oas oas={oas} operations={operations} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        {templates.operations && <Operations.File templates={templates.operations} />}
      </Oas>,
      { meta: { pluginManager, plugin } },
    )

    return root.files
  }

  async operation(operation: Operation, options: PluginOptions['resolvedOptions']): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })

    const templates = {
      operations: Operations.templates,
      client: Client.templates,
      ...options.templates,
    }

    if (!templates.client) {
      return []
    }

    root.render(
      <Oas oas={oas} operations={[operation]} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        <Oas.Operation operation={operation}>
          <Client.File templates={templates.client} />
        </Oas.Operation>
      </Oas>,
      { meta: { pluginManager, plugin: { ...plugin, options } } },
    )

    return root.files
  }

  async get(): OperationMethodResult<FileMeta> {
    return null
  }

  async post(): OperationMethodResult<FileMeta> {
    return null
  }

  async put(): OperationMethodResult<FileMeta> {
    return null
  }
  async patch(): OperationMethodResult<FileMeta> {
    return null
  }
  async delete(): OperationMethodResult<FileMeta> {
    return null
  }
}
