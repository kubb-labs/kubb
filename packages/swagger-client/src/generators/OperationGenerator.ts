import { createJSDocBlockText, getRelativePath } from '@kubb/core'
import { OperationGenerator as Generator, getComments, getParams, Path } from '@kubb/swagger'
import { pluginName as swaggerTypescriptPluginName } from '@kubb/swagger-ts'

import { pluginName } from '../plugin.ts'

import type { File, OptionalPath, PluginContext } from '@kubb/core'
import type { HttpMethod, Oas, Operation, OperationSchemas, Resolver } from '@kubb/swagger'
import type { ResolvePathOptions } from '../types.ts'

type Options = {
  clientPath?: OptionalPath
  oas: Oas
  resolvePath: PluginContext<ResolvePathOptions>['resolvePath']
  resolveName: PluginContext['resolveName']
}

export class OperationGenerator extends Generator<Options> {
  resolve(operation: Operation): Resolver {
    const { resolvePath, resolveName } = this.options

    const name = resolveName({ name: operation.getOperationId(), pluginName })

    if (!name) {
      throw new Error('Name should be defined')
    }

    const fileName = `${name}.ts`
    const filePath = resolvePath({
      fileName,
      options: { tag: operation.getTags()[0]?.name },
    })

    if (!filePath) {
      throw new Error('Filepath should be defined')
    }

    return {
      name,
      fileName,
      filePath,
    }
  }

  resolveType(operation: Operation): Resolver {
    const { resolvePath, resolveName } = this.options

    const name = resolveName({ name: operation.getOperationId(), pluginName: swaggerTypescriptPluginName })

    if (!name) {
      throw new Error('Name should be defined')
    }

    const fileName = `${name}.ts`
    const filePath = resolvePath({ fileName, options: { tag: operation.getTags()[0]?.name }, pluginName: swaggerTypescriptPluginName })

    if (!filePath) {
      throw new Error('Filepath should be defined')
    }

    return {
      name,
      fileName,
      filePath,
    }
  }

  async all(paths: Record<string, Record<HttpMethod, Operation>>): Promise<File | null> {
    const { resolvePath, resolveName, oas } = this.options

    const controllerName = resolveName({ name: 'operations' })

    if (!controllerName) {
      throw new Error('controllerName should be defined')
    }

    const controllerId = `${controllerName}.ts`
    const controllerFilePath = resolvePath({
      fileName: controllerId,
    })

    if (!controllerFilePath) {
      return null
    }
    // end controller setup

    const sources: string[] = []

    const groupedByOperationId: Record<string, { path: string; method: HttpMethod }> = {}

    Object.keys(paths).forEach((path) => {
      Object.keys(paths[path]).forEach((method) => {
        const operation = oas.operation(path, method as HttpMethod)
        if (operation) {
          groupedByOperationId[operation.getOperationId()] = {
            path: new Path(path).URL,
            method: method as HttpMethod,
          }
        }
      })
    })

    sources.push(`export const operations = ${JSON.stringify(groupedByOperationId)} as const;`)

    return {
      path: controllerFilePath,
      fileName: controllerId,
      source: sources.join('\n'),
      imports: [],
    }
  }

  async get(operation: Operation, schemas: OperationSchemas): Promise<File | null> {
    const { clientPath } = this.options

    const controller = this.resolve(operation)
    const type = this.resolveType(operation)

    const comments = getComments(operation)
    const sources: string[] = []
    const pathParamsTyped = getParams(schemas.pathParams, { typed: true })

    const generics = [`TData = ${schemas.response.name}`].filter(Boolean)
    const clientGenerics = ['TData'].filter(Boolean)
    const params = [pathParamsTyped, schemas.queryParams?.name ? `params?: ${schemas.queryParams?.name}` : ''].filter(Boolean)

    sources.push(`
      ${createJSDocBlockText({ comments })}
      export function ${controller.name} <${generics.join(', ')}>(${params.join(', ')}) {
        return client<${clientGenerics.join(', ')}>({
          method: "get",
          url: ${new Path(operation.path).template},
          ${schemas.queryParams?.name ? 'params,' : ''}
        });
      };
    `)

    return {
      path: controller.filePath,
      fileName: controller.fileName,
      source: sources.join('\n'),
      imports: [
        {
          name: 'client',
          path: clientPath ? getRelativePath(controller.filePath, clientPath) : '@kubb/swagger-client/client',
        },
        {
          name: [schemas.response.name, schemas.pathParams?.name, schemas.queryParams?.name].filter(Boolean),
          path: getRelativePath(controller.filePath, type.filePath),
          isTypeOnly: true,
        },
      ],
    }
  }

  async post(operation: Operation, schemas: OperationSchemas): Promise<File | null> {
    const { clientPath } = this.options

    const controller = this.resolve(operation)
    const type = this.resolveType(operation)

    const comments = getComments(operation)
    const sources: string[] = []
    const pathParamsTyped = getParams(schemas.pathParams, { typed: true })

    const generics = [`TData = ${schemas.response.name}`, schemas.request?.name ? `TVariables = ${schemas.request?.name}` : ''].filter(Boolean)
    const clientGenerics = ['TData', schemas.request?.name ? 'TVariables' : ''].filter(Boolean)
    const params = [
      pathParamsTyped,
      schemas.request?.name ? 'data: TVariables' : '',
      schemas.queryParams?.name ? `params?: ${schemas.queryParams?.name}` : '',
    ].filter(Boolean)

    sources.push(`
      ${createJSDocBlockText({ comments })}
      export function ${controller.name} <${generics.join(', ')}>(${params.join(', ')}) {
        return client<${clientGenerics.join(', ')}>({
          method: "post",
          url: ${new Path(operation.path).template},
          ${schemas.request?.name ? 'data,' : ''}
          ${schemas.queryParams?.name ? 'params,' : ''}
        });
      };
    `)

    return {
      path: controller.filePath,
      fileName: controller.fileName,
      source: sources.join('\n'),
      imports: [
        {
          name: 'client',
          path: clientPath ? getRelativePath(controller.filePath, clientPath) : '@kubb/swagger-client/client',
        },
        {
          name: [schemas.request?.name, schemas.response.name, schemas.pathParams?.name, schemas.queryParams?.name].filter(Boolean),
          path: getRelativePath(controller.filePath, type.filePath),
          isTypeOnly: true,
        },
      ],
    }
  }

  async put(operation: Operation, schemas: OperationSchemas): Promise<File | null> {
    const { clientPath } = this.options

    const controller = this.resolve(operation)
    const type = this.resolveType(operation)

    const comments = getComments(operation)
    const sources: string[] = []
    const pathParamsTyped = getParams(schemas.pathParams, { typed: true })

    const generics = [`TData = ${schemas.response.name}`, schemas.request?.name ? `TVariables = ${schemas.request?.name}` : ''].filter(Boolean)
    const clientGenerics = ['TData', schemas.request?.name ? 'TVariables' : ''].filter(Boolean)
    const params = [
      pathParamsTyped,
      schemas.request?.name ? 'data: TVariables' : '',
      schemas.queryParams?.name ? `params?: ${schemas.queryParams?.name}` : '',
    ].filter(Boolean)

    sources.push(`
      ${createJSDocBlockText({ comments })}
      export function ${controller.name} <${generics.join(', ')}>(${params.join(', ')}) {
        return client<${clientGenerics.join(', ')}>({
          method: "put",
          url: ${new Path(operation.path).template},
          ${schemas.request?.name ? 'data,' : ''}
          ${schemas.queryParams?.name ? 'params,' : ''}
        });
      };
    `)

    return {
      path: controller.filePath,
      fileName: controller.fileName,
      source: sources.join('\n'),
      imports: [
        {
          name: 'client',
          path: clientPath ? getRelativePath(controller.filePath, clientPath) : '@kubb/swagger-client/client',
        },
        {
          name: [schemas.request?.name, schemas.response.name, schemas.pathParams?.name, schemas.queryParams?.name].filter(Boolean),
          path: getRelativePath(controller.filePath, type.filePath),
          isTypeOnly: true,
        },
      ],
    }
  }

  async delete(operation: Operation, schemas: OperationSchemas): Promise<File | null> {
    const { clientPath } = this.options

    const controller = this.resolve(operation)
    const type = this.resolveType(operation)

    const comments = getComments(operation)
    const sources: string[] = []
    const pathParamsTyped = getParams(schemas.pathParams, { typed: true })

    const generics = [`TData = ${schemas.response.name}`, schemas.request?.name ? `TVariables = ${schemas.request?.name}` : ''].filter(Boolean)
    const clientGenerics = ['TData', schemas.request?.name ? 'TVariables' : ''].filter(Boolean)
    const params = [
      pathParamsTyped,
      schemas.request?.name ? 'data: TVariables' : '',
      schemas.queryParams?.name ? `params?: ${schemas.queryParams?.name}` : '',
    ].filter(Boolean)

    sources.push(`
      ${createJSDocBlockText({ comments })}
      export function ${controller.name} <${generics.join(', ')}>(${params.join(', ')}) {
        return client<${clientGenerics.join(', ')}>({
          method: "delete",
          url: ${new Path(operation.path).template},
          ${schemas.request?.name ? 'data,' : ''}
          ${schemas.queryParams?.name ? 'params,' : ''}
        });
      };
    `)

    return {
      path: controller.filePath,
      fileName: controller.fileName,
      source: sources.join('\n'),
      imports: [
        {
          name: 'client',
          path: clientPath ? getRelativePath(controller.filePath, clientPath) : '@kubb/swagger-client/client',
        },
        {
          name: [schemas.request?.name, schemas.response.name, schemas.pathParams?.name, schemas.queryParams?.name].filter(Boolean),
          path: getRelativePath(controller.filePath, type.filePath),
          isTypeOnly: true,
        },
      ],
    }
  }
}
