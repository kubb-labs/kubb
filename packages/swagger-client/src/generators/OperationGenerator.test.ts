import { OasManager } from '@kubb/swagger'

import { format } from '../../mocks/format.ts'
import { OperationGenerator } from './OperationGenerator.ts'

import type { KubbPlugin, PluginContext, PluginManager } from '@kubb/core'
import type { GetOperationGeneratorOptions } from '@kubb/swagger'

describe('OperationGenerator', () => {
  const resolvePath = () => './pets.ts'
  const resolveName: PluginContext['resolveName'] = ({ name }) => name

  test('[GET] should generate code based on a pathParamsType `inline`', async () => {
    const oas = await OasManager.parseFromConfig({
      root: './',
      output: { path: 'test', clean: true },
      input: { path: 'packages/swagger-client/mocks/petStore.yaml' },
    })

    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      dataReturnType: 'data',
      pathParamsType: 'inline',
    }

    const og = await new OperationGenerator(
      options,
      {
        oas,
        skipBy: [],
        pluginManager: { resolvePath, resolveName } as unknown as PluginManager,
        plugin: {} as KubbPlugin,
        contentType: undefined,
      },
    )
    const operation = oas.operation('/pets/{pet_id}', 'get')

    const get = await og.get(operation, og.getSchemas(operation), {} as typeof og.options)

    expect(await format(get?.source)).toMatchSnapshot()
  })

  test('[GET] should generate code based on a pathParamsType `object`', async () => {
    const oas = await OasManager.parseFromConfig({
      root: './',
      output: { path: 'test', clean: true },
      input: { path: 'packages/swagger-client/mocks/petStore.yaml' },
    })

    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      dataReturnType: 'data',
      pathParamsType: 'object',
    }

    const og = await new OperationGenerator(
      options,
      {
        oas,
        skipBy: [],
        pluginManager: { resolvePath, resolveName } as unknown as PluginManager,
        plugin: {} as KubbPlugin,
        contentType: undefined,
      },
    )
    const operation = oas.operation('/pets/{pet_id}', 'get')

    const get = await og.get(operation, og.getSchemas(operation), options)

    expect(await format(get?.source)).toMatchSnapshot()
  })
})
