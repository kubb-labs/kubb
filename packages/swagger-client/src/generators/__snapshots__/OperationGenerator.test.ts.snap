// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OperationGenerator > [GET] should generate code based on a pathParamsType \`inline\` 1`] = `
"/**
 * @summary Info for a specific pet
 * @link /pets/:pet_id
 */
export async function showPetById<TData = ShowPetByIdQueryResponse>(
  petId: ShowPetByIdPathParams['pet_id'],
  testId: ShowPetByIdPathParams['testId'],
  options: Partial<Parameters<typeof client>[0]> = {},
): Promise<ResponseConfig<TData>> {
  const { data: resData } = await client<TData>({
    method: 'get',
    url: \`/pets/\${petId}\`,
    ...options,
  })

  return resData
}
"
`;

exports[`OperationGenerator > [GET] should generate code based on a pathParamsType \`object\` 1`] = `
"/**
 * @summary Info for a specific pet
 * @link /pets/:pet_id
 */
export async function showPetById<TData = ShowPetByIdQueryResponse>(
  { petId, testId }: ShowPetByIdPathParams,
  options: Partial<Parameters<typeof client>[0]> = {},
): Promise<ResponseConfig<TData>['data']> {
  const { data: resData } = await client<TData>({
    method: 'get',
    url: \`/pets/\${petId}\`,
    ...options,
  })

  return resData
}
"
`;
